feat(langchain): allow configuration of metadata key prefix

ianchi opened the pull request:
<!-- Thanks for submitting a PR! To make sure this gets merged quickly, make sure to check the following checkboxes. -->

Allowing to configure the key prefix for metadata enables use on other backends.
As an example for Langfuse v3 otel backend, you can use prefix `langfuse.metadata` to make user.id and session.id available.

- [ ] I have added tests that cover my changes.
- [ ] If adding a new instrumentation or changing an existing one, I've added screenshots from some observability platform showing the change.
- [ ] PR name follows conventional commits format: `feat(instrumentation): ...` or `fix(instrumentation): ...`.
- [ ] (If applicable) I have updated the documentation accordingly.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/traceloop/openllmetry?pullRequest=3367) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/traceloop/openllmetry?pullRequest=3367) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/traceloop/openllmetry?pullRequest=3367) it.</sub>

nirga commented on packages/opentelemetry-instrumentation-langchain/opentelemetry/instrumentation/langchain/__init__.py:
```diff
@@ -37,10 +37,24 @@ def __init__(
         exception_logger=None,
         disable_trace_context_propagation=False,
         use_legacy_attributes: bool = True,
+        metadata_key_prefix: Optional[str] = None
```
this should default to the way it works tdoay

ianchi replied:
It defaulted from Config, but I also added here for clarity

ianchi commented on packages/opentelemetry-instrumentation-langchain/opentelemetry/instrumentation/langchain/callback_handler.py:
```diff
@@ -290,7 +290,7 @@ def _create_span(
                 _set_span_attribute(
                     span,
-                    f"{SpanAttributes.TRACELOOP_ASSOCIATION_PROPERTIES}.{key}",
+                    f"{Config.metadata_key_prefix}.{key}",
```
fixed

nirga commented on packages/opentelemetry-instrumentation-langchain/opentelemetry/instrumentation/langchain/__init__.py:
```diff
@@ -37,10 +36,24 @@ def __init__(
+                used to log exceptions that occur during instrumentation. If None, exceptions will not be logged.
+            disable_trace_context_propagation: If True, disables trace context propagation to LLM providers.
+            use_legacy_attributes: If True, uses span attributes for Inputs/Outputs instead of events.
+            metadata_key_prefix: Prefix for metadata keys added to spans. If None, defaults to
```
Nit: this is inaccurate - only if not set the default kicks in. If it's None then it's None

nirga commented on packages/opentelemetry-instrumentation-langchain/opentelemetry/instrumentation/langchain/__init__.py:
```diff
@@ -37,10 +36,24 @@ def __init__(
         exception_logger=None,
         disable_trace_context_propagation=False,
         use_legacy_attributes: bool = True,
+        metadata_key_prefix: Optional[str] = SpanAttributes.TRACELOOP_ASSOCIATION_PROPERTIES
```
Type shouldn't be optional - it always has a value

nirga reviewed (commented):
Thanks @ianchi - left a small comment

nirga reviewed (commented):
Thanks @ianchi - CI seems to fail. Can you take a look?

ianchi commented:
@nirga  I see that commit d9a0d7d00561ae920190cf3cd894491bdac036c1 tried to address part of the problem, but it introduced yet another bug. It is using:

`SpanAttributes.GEN_AI_USAGE_CACHE_CREATION_INPUT_TOKENS,`

but this does not exist in SpanAttributes.

In this PR I rebased and fixed it for langchain, but that PR also introduces the bug in Anthropic
